### PR TITLE
Define core version at create command and defines an update command

### PR DESCRIPTION
**Pull request by gabrielbussolo**
# Summary of this proposal

Define core version at create command and defines an update command

https://user-images.githubusercontent.com/4117683/178740914-4a8bbea6-2233-4b63-9fe9-3dccfe5c412c.mov

## Issue(s) number(s)

*Disclaimer: Please do not create a Pull Request without creating an issue first.*

## Checklist for submitter

- [ ] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [ ] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

Linked to #81 

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

**Review by niedbalski — CHANGES_REQUESTED** [review]
1. Lets use version "github.com/hashicorp/go-version" for semver validator (look at https://github.com/calyptia/cloud/blob/bffce97b74aa39fdf046e62e75a98c7990d5ee4c/aggregator.go#L99) to standardise across. the projects.
2. The update command parameter should be called --version (not --new-version) as that's not the attribute name you are trying to update.

File: cmd/calyptia/update.go [hunk 1]
```diff
@@ -5,17 +5,18 @@ import "github.com/spf13/cobra"
...
 		Use:   "update",
-		Short: "Update aggregators, pipelines, etc.",
+		Short: "Update aggregators, pipelines, core instances etc.",
 	}
 
 	cmd.AddCommand(
 		newCmdUpdateProject(config),
 		newCmdUpdateAgent(config),
-		newCmdUpdateAggregator(config),
+		//newCmdUpdateAggregator(config),
```
Comment (niedbalski) [inline cmd/calyptia/update.go:14]: don't comment this out for now.

File: k8s/client.go [hunk 2]
```diff
@@ -229,3 +230,41 @@ func (client *Client) DeleteSecretByLabel(ctx context.Context, label, ns string)
...
+	return err
+}
+func (client *Client) FindDeploymentByName(ctx context.Context, name string) (*appsv1.Deployment, error) {
+	deployment, err := client.AppsV1().Deployments(client.Namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return deployment, nil
+}
+func ValidSemVer(instanceVersion string) (bool, error) {
```
Comment (niedbalski) [inline k8s/client.go:257]: check my comment on the validator.

**Comment by niedbalski** [comment]
@gabrielbussolo For the index validator (https://raw.githubusercontent.com/calyptia/core-images-index/main/container.index.json)

File: cmd/calyptia/update.go (hunk 1 above)
Comment (gabrielbussolo) [reply cmd/calyptia/update.go:14]: this guy actually is core_instance command that we were using "aggregator" before, i transfered the logic to the new comand

**Comment by niedbalski** [comment]
@gabrielbussolo all good with the index?

File: cmd/calyptia/update_core_instance_k8s.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	cloud "github.com/calyptia/api/types"
+	"github.com/calyptia/cli/k8s"
+	"github.com/spf13/cobra"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"net/http"
+)
+
+const calyptiaCoreVersions = "https://raw.githubusercontent.com/calyptia/core-images-index/main/container.index.json"
```
Comment (niedbalski) [inline cmd/calyptia/update_core_instance_k8s.go:16]: rename to calyptiaCoreImageIndexURL

File: cmd/calyptia/update_core_instance_k8s.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	return cmd
+}
+
+func VerifyCoreVersion(newVersion string) error {
+	if newVersion == "" {
+		return nil
+	}
+	availableVersions, err := getAvailableVersions()
+	if err != nil {
+		return err
```
Comment (niedbalski) [inline cmd/calyptia/update_core_instance_k8s.go:114]: expand this error to +               return fmt.Errorf("could not fetch available versions of calyptia-core: %w", err)

File: cmd/calyptia/update_core_instance_k8s.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		},
+	}
+	fs := cmd.Flags()
+	fs.StringVar(&newVersion, "version", "", "New version of the calyptia-core instance")
+	fs.StringVar(&newName, "name", "", "New core instance name")
+	clientcmd.BindOverrideFlags(configOverrides, fs, clientcmd.RecommendedConfigOverrideFlags("kube-"))
+	return cmd
+}
+
+func VerifyCoreVersion(newVersion string) error {
```
Comment (niedbalski) [inline cmd/calyptia/update_core_instance_k8s.go:108]: verifyCoreImageVersion -- Add unit tests for this function (as is exported). (cover the 2 cases - good/bad request to the index).

File: cmd/calyptia/update_core_instance_k8s.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+		return err
+	}
+	if !contains(availableVersions, newVersion) {
+		return fmt.Errorf("version %s is not available", newVersion)
+	}
+
+	return nil
+}
+
+func getAvailableVersions() ([]string, error) {
```
Comment (niedbalski) [inline cmd/calyptia/update_core_instance_k8s.go:123]: getCoreImageTags or getCoreImageVersions()

File: k8s/client.go [hunk 7]
```diff
@@ -229,3 +228,20 @@ func (client *Client) DeleteSecretByLabel(ctx context.Context, label, ns string)
...
+
+func (client *Client) UpdateDeployment(ctx context.Context, name, image, version string) error {
+	deployment, err := client.FindDeploymentByName(ctx, name)
+	if err != nil {
+		return err
+	}
+	deployment.Spec.Template.Spec.Containers[0].Image = fmt.Sprintf("%s:%s", image, version)
+	_, err = client.AppsV1().Deployments(client.Namespace).Update(ctx, deployment, metav1.UpdateOptions{})
+	return err
+}
```
Comment (niedbalski) [inline k8s/client.go:240]: line break

**Review by niedbalski — CHANGES_REQUESTED** [review]
1. Generally, Add unit tests for the method VerifyCoreImageVersion.
2. Minor nitpicks in the code.

**Review by niedbalski — APPROVED** [review]

**Comment by niedbalski** [comment]
@gabrielbussolo 

➜ cli (update-version) ✔ ./calyptia update core_instance kubernetes core --version v0.2.1
calyptia-core instance "core" was successfully updated to version v0.2.1

➜ cli (update-version) ✔ ./calyptia update core_instance kubernetes core --version v0.2.6
Error: version v0.2.6 is not available

**Pull request merged**